### PR TITLE
Fix progress functionality for recursive merges

### PR DESCRIPTION
**Pull request by acwhite211**
Fix error so that the total is updated during a recursive merge

### CODE REVIEW

**Review by CarolineDenis — APPROVED** [review]

**Review by carlosmbe — CHANGES_REQUESTED** [review]
It works better than before. No longer negative. However, time remaining only escalates as time goes on. Here is a small snippet 

https://github.com/specify/specify7/assets/53784701/c541277d-5ab4-4ddd-9c5c-0a1c388ac849

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx [hunk 1]
```diff
@@ -99,10 +100,7 @@ export function Status({
           {state.status === 'MERGING' && (
             <>
               <Progress max={state.total} value={state.current} />
-              <RemainingLoadingTime
-                current={state.current}
-                total={state.total}
-              />
+              <p>{percentage}%</p>
```
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/Merging/Status.tsx:103]: hmm, I don't like this change.
the video Carlos sent looked like a bug - if so, it needs to be fixed rather than the feature removed

Comment (CarolineDenis) [reply specifyweb/frontend/js_src/lib/components/Merging/Status.tsx:103]: Changing the remaining time by a percentage is what as been decided during the Sp meeting with the all team

**Review by carlosmbe — CHANGES_REQUESTED** [review]
Nicely done.

Just a few aesthetic related things. Wouldn't call them road blocks. I lean towards polish. 

Firstly, change the highest number from 100% to 99% As right now it pauses for a moment when at 100. I believe having it stop at 99% would be ideal as the user can assume that it hit 100% when the merge success message shows up.

Secondly, can we change the red exclamation mark on top to something else. The tester deep inside me senses danger when they see a big red !!!!!

Also, at `Merge Successful`, we should make the close button Green not red.

Thanks for the work. 

https://github.com/specify/specify7/assets/53784701/d274ae38-30fb-4529-92cb-c0ee5767fa10

**Comment by CarolineDenis** [comment]
> Nicely done.
> 
> Just a few aesthetic related things. Wouldn't call them road blocks. I lean towards polish.
> 
> Firstly, change the highest number from 100% to 99% As right now it pauses for a moment when at 100. I believe having it stop at 99% would be ideal as the user can assume that it hit 100% when the merge success message shows up.
> 
> Secondly, can we change the red exclamation mark on top to something else. The tester deep inside me senses danger when they see a big red !!!!!
> 
> Also, at `Merge Successful`, we should make the close button Green not red.
> 
> Thanks for the work.
> 
>  Screen.Recording.2023-08-18.at.11.01.38.AM.mp4

For: > Secondly, can we change the red exclamation mark on top to something else. The tester deep inside me senses danger when they see a big red !!!!!
@carlosmbe, any suggestion?

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx (hunk 1 above)
Comment (melton-jason) [reply specifyweb/frontend/js_src/lib/components/Merging/Status.tsx]: The thought process behind the change is that the time remaining is often sporadic by nature when it comes to merging. 
 The current and total in the requests have a tendency to "jump around" a lot. More specifically, whenever the merging process needs to merge records recursively, the 'total' will need to be updated to reflect the number of added columns, which may cause the `time remaining` function to fluctuate. 

The thought process was that using a percentage "handles" the visual turbulence better from a user standpoint, since it is not _heavily_ changing the progress UI based on a drastic change in the current and total. 
In other words, a percentage is hopefully smoother. 
(Not my own personal thoughts here, just what I picked up as the reasoning behind the change)

> the video Carlos sent looked like a bug - if so, it needs to be fixed rather than the feature removed

While it may be tricky to have the time remaining function handle results like how it might get them form agent merging, I agree this is a bug and should probably be looked into/fixed

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx [hunk 2]
```diff
@@ -94,6 +95,15 @@ export function Status({
       dimensionsKey="merging-progress"
       header={statusLocalization[state.status]}
       onClose={undefined}
+      icon={
+        state.status === 'PENDING' ? (
+          icons.cog
+        ) : state.status === 'MERGING' ? (
+          <span className="text-red-500">{icons.exclamationCircle}</span>
+        ) : (
+          <span className="text-green-500">{icons.checkCircle}</span>
```
Comment (melton-jason) [inline specifyweb/frontend/js_src/lib/components/Merging/Status.tsx:104]: Not the biggest fan of hardcoding the colors here. There are considerations with light/dark mode, and accessibility concerns to those that are color blind. 
If the colors are important, can you use a `dialogIcon` then, to stay consistent with the rest of Specify?
https://github.com/specify/specify7/blob/2a259ca22b439ba71b9313705cb2a253ac8edd6e/specifyweb/frontend/js_src/lib/components/Atoms/Icons.tsx#L111-L124

**Review by melton-jason — CHANGES_REQUESTED** [review]

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx (hunk 1 above)
Comment (maxxxxxdlp) [reply specifyweb/frontend/js_src/lib/components/Merging/Status.tsx]: makes sense.

this is why the progress bars in any software with good UX are mostly fake.
they just gradually move across, and get slower the closer they reach to the end, and get faster the more actual progress is completed - until finally when everything is actually complete, instead of disappearing right away (which would break the illusion), the scroll bar reaches the end in a span of 1-2 seconds - giving immense satisfaction to the user

macOS update progress bar might be the best example of this
though, windows have also gotten much better with their progress bars since windows xp days

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx (hunk 2 above)
Comment (maxxxxxdlp) [reply specifyweb/frontend/js_src/lib/components/Merging/Status.tsx:104]: @melton-jason great suggestion!

the only thing I can say in defense is that the `-500` colors are the only colors that work fine both in light and dark mode (though fine does not mean ideal)

**Review by melton-jason — APPROVED** [review]

**Review by carlosmbe — APPROVED** [review]

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx [hunk 3]
```diff
@@ -40,13 +43,18 @@ export function Status({
...
         // eslint-disable-next-line @typescript-eslint/naming-convention
         headers: { Accept: 'application/json' },
       })
         .then(
           ({
-            data: { taskstatus: taskStatus, taskprogress: taskProgress },
+            data: {
+              taskstatus: taskStatus,
+              taskprogress: taskProgress,
+              response: response,
```
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/Merging/Status.tsx:56]:
```suggestion
              response,
```

File: specifyweb/specify/views.py [hunk 4]
```diff
@@ -965,7 +966,8 @@ def merging_status(request, merge_id: int) -> http.HttpResponse:
     status = {
         'taskstatus': task_status,
         'taskprogress': task_progress,
-        'taskid': merge_id
+        'taskid': merge_id,
+        'response': merge.response
```
Comment (maxxxxxdlp) [inline specifyweb/specify/views.py:970]: (optional) does `response` always contain either error message or null? in such case, consider renaming it

**Review by maxxxxxdlp — APPROVED** [review]

File: specifyweb/specify/views.py (hunk 4 above)
Comment (CarolineDenis) [reply specifyweb/specify/views.py:970]: it's either a string or an empty string

**Review by bronwyncombs — APPROVED** [review]
Functioning well thank you

https://github.com/specify/specify7/assets/135047322/c6e4d570-868a-4150-a941-5dc7ed98ba85

File: specifyweb/frontend/js_src/lib/components/Merging/Status.tsx [hunk 5]
```diff
@@ -67,10 +78,29 @@ export function Status({
...
     <Dialog
       buttons={
-        state.status === 'MERGING' ? (
+        state.status === 'FAILED' ? (
+          <>
+            <Button.Info
+              onClick={(): void =>
+                void downloadFile(
+                  `Merging ${mergingId} Crash Report - ${new Date().toJSON()}.txt`,
+                  responseError
```
Comment (maxxxxxdlp) [inline specifyweb/frontend/js_src/lib/components/Merging/Status.tsx:92]: (optional) replace `responseError` with `produceStackTrace(responseError)` so that the file you would receive from the user contains much more useful information (including, which records the user was trying to merge)

**Review by maxxxxxdlp — APPROVED** [review]

File: specifyweb/specify/views.py (hunk 4 above)
Comment (maxxxxxdlp) [reply specifyweb/specify/views.py]: well, and when it's a string, it is an error string, right?
so wouldn't `errorMessage` be more appropriate?

also, try to not use empty strings in API responses - it should be explicitly `null` or the error message.

**Review by melton-jason — APPROVED** [review]
